### PR TITLE
[LTC] Auto-gen _log_softmax for TorchScript backend in LTC

### PR DESCRIPTION
**Pull request by nkreeger**
This PR introduces some changes needed to the generated code for PyTorch ops that begin with `_`. This first change introduces the `_log_softmax` op with proposed changes. All unit tests inside `test/cpp/build/test_ptltc` continue to pass.

It wasn't clear what the best naming and handling convention is with operators that have a high-level definition, but a `_` prefix. I noticed that the tensor.h/tensor_methods.cpp exposes a "log_softmax" without the prefix. 

Happy to make any changes that aren't wanted - just looking for some feedback if I'm making the right change here :-)

### CODE REVIEW

**Comment by nkreeger** [comment]
Note - the new class in `LazyLazyIr.h` will read as following:

```c++
class LogSoftmax : public Node {
 public:
  LogSoftmax(ir::Value self, int64_t dim, bool half_to_float, at::ScalarType out_dtype, std::vector<int64_t> out_shape)
      : Node(ir::OpKind(at::aten::_log_softmax),
              {self},
              /*shape=*/lazy_tensors::Shape(out_dtype, out_shape),
              /*num_outputs=*/1,
              lazy_tensors::util::MHash(dim, half_to_float)),
        out_dtype_(out_dtype),
        out_shape_(out_shape),

        dim_(dim),
        half_to_float_(half_to_float)

  {
    //  throw std::runtime_error("need to hash scalars properly");
  }

  std::string ToString() const override {
    std::stringstream ss;
    ss << Node::ToString();
    lazy_tensors::ToString("dim", dim_, ss);
    lazy_tensors::ToString("half_to_float", half_to_float_, ss);
    return ss.str();
  }

  NodePtr Clone(OpList operands) const override {
      ir::MakeNode<ir::ops::LogSoftmax>(operands.at(0),dim_,half_to_float_,out_dtype_,out_shape_);
  }

  c10::ScalarType out_dtype_;
  std::vector<int64_t> out_shape_;
  int64_t dim_;
  bool half_to_float_;
};
```

File: tools/codegen/api/lazy.py [hunk 1]
```diff
@@ -118,7 +122,8 @@ def __init__(self, func: FunctionSchema):
 
     @property
     def node_name(self) -> str:
-        return str(self.name).lower().capitalize()
+        op_name = str(self.name).lower()
+        return ''.join(word.capitalize() or '' for word in op_name.split('_'))
```
Comment (wconstab) [inline tools/codegen/api/lazy.py:126]: what's the `or ''` for?  These seem the same to me, but i'm probably missing some case:

```
>>> [word.capitalize() or '' for word in "_log_softmax".split("_")]
['', 'Log', 'Softmax']
>>> [word.capitalize() for word in "_log_softmax".split("_")]
['', 'Log', 'Softmax']
```

**Review by wconstab — APPROVED** [review]
LGTM overall! Thanks @nkreeger!

Regarding the naming- ops preceded* by underscore, I believe, are just a signal that they are not meant to be user facing functions (e.g. used by other user-facing functions as internal impl helper).  @ezyang could you confirm?

*Not to be confused with ops with names _followed_ by an underscore, which has a clear meaning that the op operates in-place on one of its inputs rather than being functional and returning a new tensor.
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/README.md
^ most of this stuff is spelled out there, except I couldn't find anything about preceding underscore.

One thing I probably ignored in the codegen design is what to do if we support two variants of the same op.  (Imagine we support both _log_softmax and log_softmax.Dimname variants).  The two directions we could go are 
A) make a new IR class for each one and name it uniquely (we don't name uniquely today)
B) share an IR class for each 'base' op name, such as 'log_softmax'

(A) seems simplest actually, and maybe that's my leaning.  (B) is closest to what's in the current codebase, which makes sense since it saves manual work of writing more classes.  But if we go to (A) it also impacts XLA having to lower more specific IR node classes to HLO ,  since they have to write manual lowerings while we codegen TS lowerings.

For now let's land your code as-is and we'll come back to this design issue.

File: tools/codegen/api/lazy.py [hunk 2]
```diff
@@ -34,6 +34,10 @@ def process_ir_type(typ: Type) -> Union[BaseCType, VectorCType, OptionalCType, L
             return BaseCType(valueT)
         elif typ.name == BaseTy.Scalar:
             return BaseCType(scalarT)
+        elif typ.name == BaseTy.int:
+            return BaseCType(intT)
+        elif typ.name == BaseTy.bool:
```
Comment (alanwaketan) [inline tools/codegen/api/lazy.py:39]: I added this in one of my local patch too. haha. I will rebase on your patch then.

File: tools/codegen/api/lazy.py (hunk 1 above)
Comment (nkreeger) [reply tools/codegen/api/lazy.py:126]: The `''` replaces any occurrences so we get `LogSoftmax` instead of `_LogSoftmax`.

**Review by alanwaketan — APPROVED** [review]
LGTM. Merging then.

**Pull request merged**